maitake/src/wait/map.rs

```diff
@@ -0,0 +1,926 @@
+                        // the queue is `Empty`, transition to `Waiting`
+                        State::Empty => match queue.compare_exchange(queue_state, State::Waiting) {
+                            Ok(_) => break 'to_waiting,
+                            Err(actual) => queue_state = actual,
+                        },
+                        // the queue is already `Waiting`
+                        State::Waiting => break 'to_waiting,
+                        State::Closed => return closed(),
+                    }
+                }
+
+                // enqueue the node
+                *this.state = WaitState::Waiting;
+                this.node.as_mut().with_mut(|node| {
+                    unsafe {
+                        // safety: we may mutate the node because we are
+                        // holding the lock.
+                        (*node).waker = Wakeup::Waiting(cx.waker().clone());
+                    }
+                });
+                let ptr = unsafe { NonNull::from(Pin::into_inner_unchecked(self)) };
+                waiters.push_front(ptr);
+
+                Poll::Pending
+            }
+            WaitState::Waiting => {
+                let mut _waiters = queue.queue.lock();
+                this.node.with_mut(|node| unsafe {
+                    // safety: we may mutate the node because we are
+                    // holding the lock.
+                    let node = &mut *node;
+                    match node.waker {
+                        // We already had a waker, but are now getting another one.
+                        // Store the new one, droping the old one
+                        Wakeup::Waiting(ref mut waker) => {
+                            if !waker.will_wake(cx.waker()) {
+                                *waker = cx.waker().clone();
+                            }
+                            Poll::Pending
+                        }
+                        // We have received the data, take the data out of the
+                        // future, and provide it to the poller
+                        Wakeup::DataReceived => {
+                            *this.state = WaitState::Completed;
+                            node.waker = Wakeup::Retreived;
+
+                            let val = this.val.with_mut(|v| {
+                                let mut retval = MaybeUninit::<V>::uninit();
+                                (*v).as_mut_ptr()
+                                    .copy_to_nonoverlapping(retval.as_mut_ptr(), 1);
+                                retval.assume_init()
+                            });
+
+                            notified(val)
+                        }
+                        Wakeup::Retreived => {
+                            // TODO: Should I change the result type to also return an error
+                            // for this case? Right now error can only be `Closed`
+                            consumed()
+                        }
```

Comment:
it looks like the TODO comment here was addressed?
```suggestion
                        Wakeup::Retreived => consumed(),
```